src/context.js

```diff
@@ -633,6 +646,26 @@ export default class Context {
 		const index = {};
 		const indexVariables = {};
 
+		const binarySearch = (arr, value, comparator) => {
```

Comment:
Maybe this could live in the `utils.js` file, or perhaps a new `algorithm.js` file?